framework/src/base/Assembly.C

```diff
@@ -2975,79 +2975,90 @@ Assembly::setResidualBlock(NumericVector<Number> & residual,
 void
 Assembly::addResidual(NumericVector<Number> & residual, TagID tag_id)
 {
+  auto & tag_Re = _sub_Re[residualVectorTagIndex(tag_id)];
   const std::vector<MooseVariableFEBase *> & vars = _sys.getVariables(_tid);
   for (const auto & var : vars)
     addResidualBlock(residual,
-                     _sub_Re[tag_id][var->number()],
+                     tag_Re[var->number()],
                      var->dofIndices(),
                      var->arrayScalingFactor(),
                      var->isNodal());
 }
 
 void
-Assembly::addResidual(const std::map<TagName, TagID> & tags)
+Assembly::addResidual(const std::vector<VectorTag> & vector_tags)
 {
-  for (auto & tag : tags)
-    if (_sys.hasVector(tag.second))
-      addResidual(_sys.getVector(tag.second), tag.second);
+  for (const auto & vector_tag : vector_tags)
+  {
+    mooseAssert(vector_tag._type == Moose::VECTOR_TAG_RESIDUAL, "Non-residual tag in residual");
+    if (_sys.hasVector(vector_tag._id))
+      addResidual(_sys.getVector(vector_tag._id), vector_tag._id);
+  }
 }
 
 void
 Assembly::addResidualNeighbor(NumericVector<Number> & residual, TagID tag_id)
 {
+  auto & tag_Rn = _sub_Rn[residualVectorTagIndex(tag_id)];
```

Comment:
`residualVectorTagIndex(tag_id)` should be repalced by "`_local_id`"

E.g.,

`_vector_tags[tag_id]._local_id.`

Reply:
We may check if it is a residual tag as well